doc/filters.md

```diff
@@ -0,0 +1,188 @@
+# CKAN Query Filters spec
+
+The goal of this specification is to provide a common syntax for defining filters
+that need to be applied to queries performed. It is not tied to a particular API
+endpoint or search backend, and it aims to be simple and extensible.
+
+## Overview
+
+Filters are always provided as a dictionary or as a list of dictionaries.
+
+When provided as a dictionary, the top level keys can either be:
+
+* A field name
+* A top level operator: one of `$or` or `$and`
+
+For example:
+
+```
+"filters": {
+    "field1": "value",
+    "$or": [
+        {"field2": ["value1", "value2"]},
+        {"field3": {"gte": 9000, "lt": 1000}}
+    ]
+}
+```
+
+When provided as a list of dictionaries, these are combined with the OR operator. For instance:
+
+```
+"filters": [
+    {"field1": "value1"},
+    {"field2": "value2"},
+]
+```
+
+is interpreted as:
+
+```
+"filters": {
+  "$or": [
+    {"field1": "value1"},
+    {"field2": "value2"},
+  ]
+}
+```
+
+## Individual filters
+
+Individual filters have the form:
+
+```
+  {
+    "<field_name>": {"<operator1>" : <value1>, "<operator2>" : <value2>...}
+  }
+
+```
+
+Note that fields starting with the character `$` must be escaped with another `$` character:
+
+```
+filters: {
+    "$$some_field": "some_value"
+}
+```
+
+If the operator is omitted, the `eq` (for equality) operator is assumed:
+
+```
+  {
+    "field1": "value1"
+  }
+
+```
+
+is equivalent to:
+
+```
+  {
+    "field1": {"eq" : "value1"}
+  }
+
+```
+
+If filters contain a list of values, the `in` operator is assumed, e.g.:
+
+```
+  {
+    "field1": ["value1", "value2", "value3"]
+  }
+
+```
+
+is interpreted as:
+
+```
+  {
+    "field1": {"in": ["value1", "value2", "value3"]}
+  }
+
+```
+
+## Built-in operators
+
+The following operators are supported by default:
+
+| Notation | Operation                            |
+| -------- | ------------------------------------ |
+| `eq`     | Equal                                |
+| `gt`     | Greater than                         |
+| `gte`    | Greater than or equal                |
+| `lt`     | Lower than                           |
+| `lte`    | Lower than or equal                  |
+| `in`     | Matches one of the provided values   |
+
+TODO: Additional operators can be added by plugins
+
+## Top level operators
+
+The following default top operators can be used to combine filters:
+
+| Notation | Operation                                  |
+| -------- | ------------------------------------------ |
+| `$or`    | logical OR, any of the filters can match   |
+| `$and`   | logical AND, all of the filters must match |
+
+
+Top level operators must always contain lists of at least two filters, i.e. all of the 
+following are incorrect:
+
+```
+{
+  "$or": {"field1": "value1"},
+}
+
+{
+  "$and": [
+    {"field1": "value1"}
+  ]
+}
```

Comment:
We can allow 0- and 1-element `"$or"` and `"$and"` values if we treat them like the python `any()` and `all()` functions.

```python
>>> any([])  # same as "$or": []
False
>>> all([])  # same as "$and": []
True
```
The benefit is when search filters for API calls are generated by code that code doesn't need to special-case the 0 and 1-element cases.

Reply:
I can see how ignoring empty lists is useful, but I don't understand how considering them False or True works with the actual parsing, ?

For example, what does this parse to?

```
{
   "field1": "value1",
   "$or": []
}
```
Is it something like:
```
field1 = value AND False
```

Reply:
If the goal is make it easier for code generated filters by not chocking at empty lists so users don't have to worry about them, getting no results because you inadvertently sent  an empty list might be more confusing. You effectively need to check that you are not sending empty lists to get some results back, which defeats the purpose of not special casing empty lists right?

Reply:
Getting nothing back from an empty `$or` list is consistent.
 - `"color": ["red", "blue"]` might give some results
 - `"color": []` we would expect to give 0 results

Which is the same as:
- `"$or": [{"color": "red"}, {"color": "blue"}]` might give some results
- `"$or": []` we would expect to give 0 results

Getting 0 results is also fine because this might be a smaller part of a larger `$or` condition that we don't want to fail just because one part gives no results.

Reply:
I like the idea of treating `$and` and `$or` like `all()` and `any()` because I think it's a better interface, but if you think it causes confusion then we could raise an error in the 1-element and 0-element cases instead.